stdlib/messaging/activemq-artemis/src/main/ballerina/artemis/artemis_commons.bal

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/io;
+
+# Constant for the artemis error code.
+public const string ARTEMIS_ERROR_CODE = "{ballerina/artemis}ArtemisError";
+
+# The Artemis error record.
+#
+# + message - the error message.
+public type ArtemisError record {
+    string message = "";
+    !...;
+};
+
+# The url configuration for `Producer` and `Consumer`
+#
+# + host - The host
+# + port - The port
+# + username - The username
+# + password - The password
+public type URLConfiguration record {
+    string host;
+    int port;
+    string? username = ();
+    string? password = ();
```

Comment:
Again, should `username` and `password` be optional fields? Applicable elsewhere too.

Reply:
If we use the optional fields we need to do a lot of boiler plate error handling because of a key missing instead of being nil.  In this case it is more suitable to have a nil value instead of an optional field.

Reply:
If we use index based access, I don't see any difference. 

Say we have a record value `urlConfig` (of type `URLConfiguration`).

Currently when you say `urlConfig.username` the resultant value is of type `string?`. It will be the same if `username` is an optional field in the record and you say `urlConfig["username"]`. 

Could you explain what you meant by "boiler plate error handling" ?

With this approach we allow the user to set the `username` as `()` which is wrong IMO. I believe we should be minimizing room for user error even if we have to write a bit of extra code.

Reply:
I have fixed this in this place. But it happens to be that username and password could be null in the Artemis broker protocol which they handle. So didn't change it in the session.